src/AL_alg.jl

```diff
@@ -0,0 +1,247 @@
+export AL
+
+function AL(nlp::AbstractNLPModel, h::H, options::ROSolverOptions; kwargs...) where {H}
+  if unconstrained(nlp) || bound_constrained(nlp)
+    return AL(Val(:unc), nlp, h, options; kwargs...)
+  elseif equality_constrained(nlp)
+    return AL(Val(:equ), nlp, h, options; kwargs...)
+  else # has inequalities
+    return AL(Val(:ineq), nlp, h, options; kwargs...)
+  end
+end
+
+function AL(
+  ::Val{:unc},
+  nlp::AbstractNLPModel,
+  h::H,
+  options::ROSolverOptions;
+  subsolver = has_bounds(nlp) ? TRDH : PG,
```

Comment:
I think better choices would be TR and R2 here. R2 is typically faster than PG. TRDH will only work if h is separable.